### PR TITLE
added code for review - GA

### PR DESCRIPTION
**Pull request by julian-upc**

### CODE REVIEW

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+
+    using VertId = size_t;
+    using Facet = PointSet;
+
+    size_t n_vert;
+    std::vector<Facet> facets; // sorted by TotalOrder
+};
+
+bool operator==(CombinatorialType::PointSet a, CombinatorialType::PointSet b)
+{
```
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: Possibly use `const CombinatorialType::PointSet&` just to be explicit... maybe the compiler will do this anyway, but in any case `a` and `b` will still be valid after this function, so the compiler can't move from them

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+    return a.tsize == b.tsize && a.bits == b.bits;
+}
+bool operator!=(CombinatorialType::PointSet a, CombinatorialType::PointSet b)
```
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: ditto

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+{
+    return !(a == b);
+}
+
+
+// Obtain the line through two points
+ppl::Linear_Expression mk_line(Point p, Point q)
```
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: ditto

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+    return d2*d3*x3.get_num()*X + d1*d3*y3.get_num()*Y + d1*d2*z3.get_num();
+}
+
+// Return whether the diagram is a affine Gale diagram of a polytope.
+template <bool Neighborly>
+bool is_gale_diagram_of_polytope(const GaleDiagram& diagram)
+{
+    const auto& points = diagram.points;
+
+    for (size_t i = 0; i < points.size(); i++) {
```
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: signpost error: `i < points.size() - 1`

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+        auto p = points[i].pt;
```
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: consider `auto&`

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+        for (size_t j = i+1; j < points.size(); j++) {
+            auto q = points[j].pt;
```
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: ditto

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+        if (is_valid_face_candidate(diagram, candidate))
+            candidates.push_back(candidate);
+    }
+
+    auto implies = [](bool p, bool q) { return !p || q; };
+    result.facets.reserve(candidates.size()/2);
+
+    for (auto f : candidates) {
+        auto maximal =
+            std::all_of(candidates.begin(), candidates.end(), [=](auto c) {
```
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: You probably want `[&]` here

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+        };
+
+        if (is_valid_face_candidate(diagram, candidate))
+            candidates.push_back(candidate);
+    }
+
+    auto implies = [](bool p, bool q) { return !p || q; };
+    result.facets.reserve(candidates.size()/2);
+
+    for (auto f : candidates) {
```
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: consider `const auto& f : candidates`

**Comment by julian-upc** [comment]
Great job, guys! I love your code, so sweet & clean. I learned stuff reading it, a real pleasure! Thanks!

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp (hunk 1 above)
Comment (bielr) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: I thought that the indirections would not be worth it; `PointSet`s are just two ints so the copy should be cheaper

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp (hunk 5 above)
Comment (bielr) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: Thanks for this one! I have no idea about the internals of gmp classes, is this creating a full copy of the contents of `mpq_class` or do they use any sharing?

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp (hunk 1 above)
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: ok!

File: 2016/exercises/final_project/Gabriel-Adrian/main.cpp (hunk 5 above)
Comment (julian-upc) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: `p` and `q` are used in the call to `mk_line`, which you declared as wanting copies,
```
ppl::Linear_Expression mk_line(Point p, Point q) {
...
}
```
so yes, at some point you would end up making a copy of `points[i].pt`.

Copying `mp*_class`es should be avoided like the plague: they allocate a huge amount of memory on the heap, and the default allocators work very slowly. Treating an `mp*_class` as if it were a register (as in `register int`) and reusing them as much as possible can really help to cut down on runtime.

Comment (bielr) [inline 2016/exercises/final_project/Gabriel-Adrian/main.cpp]: That's good to know, thank you for the explaination.

**Pull request merged**